src/circuit/client.ts

```diff
@@ -0,0 +1,360 @@
+import { logger } from '@libp2p/logger'
+import { RELAY_V2_HOP_CODEC } from './multicodec.js'
+import { getExpiration, namespaceToCid } from './utils.js'
+import {
+  CIRCUIT_PROTO_CODE,
+  DEFAULT_MAX_RESERVATIONS,
+  RELAY_RENDEZVOUS_NS
+} from './constants.js'
+import type { PeerId } from '@libp2p/interface-peer-id'
+import type { AddressSorter } from '@libp2p/interface-peer-store'
+import type { Connection } from '@libp2p/interface-connection'
+import sort from 'it-sort'
+import all from 'it-all'
+import { pipe } from 'it-pipe'
+import { publicAddressesFirst } from '@libp2p/utils/address-sort'
+import { reserve } from './v2/index.js'
+import { EventEmitter, CustomEvent } from '@libp2p/interfaces/events'
+import type { Startable } from '@libp2p/interfaces/startable'
+import type { Components } from '../components.js'
+import type { RelayReservationManagerConfig } from './index.js'
+import { PeerSet, PeerMap, PeerList } from '@libp2p/peer-collections'
+
+const log = logger('libp2p:circuit:client')
+
+const noop = () => {}
+
+/**
+ * CircuitServiceInit initializes the circuit service using values
+ * from the provided config and an @type{AddressSorter}.
+ */
+export interface CircuitServiceInit extends RelayReservationManagerConfig {
+  /**
+   * Allows prioritizing addresses from the peerstore for dialing. The
+   * default behavior is to prioritise public addresses.
+   */
+  addressSorter?: AddressSorter
+  /**
+   * A callback to invoke when an error occurs in the circuit service.
+   */
+  onError?: (error: Error) => void
+}
+
+export interface RelayReservationManagerEvents {
+  'relay:reservation': CustomEvent
+}
+
+/**
+ * ReservationManager automatically makes a circuit v2 reservation on any connected
+ * peers that support the circuit v2 HOP protocol.
+ */
+export class RelayReservationManager extends EventEmitter<RelayReservationManagerEvents> implements Startable {
+  private readonly components: Components
+  private readonly addressSorter: AddressSorter
+  private readonly maxReservations: number
+  private readonly relays: PeerSet
+  private readonly reservationMap: PeerMap<ReturnType<typeof setTimeout>>
+  private readonly onError: (error: Error) => void
+  private started: boolean
+
+  constructor (components: Components, init: CircuitServiceInit) {
+    super()
+    this.started = false
+    this.components = components
+    this.addressSorter = init.addressSorter ?? publicAddressesFirst
+    this.maxReservations = init.maxReservations ?? DEFAULT_MAX_RESERVATIONS
+    this.relays = new PeerSet()
+    this.reservationMap = new PeerMap()
+    this.onError = init.onError ?? noop
+
+    this._onProtocolChange = this._onProtocolChange.bind(this)
+    this._onPeerDisconnected = this._onPeerDisconnected.bind(this)
+
+    this.components.peerStore.addEventListener('change:protocols', (evt) => {
```

Comment:
It recently surfaced that this won't trigger in case node restarts or peer reconnects as protocols won't be changed. I think we need to handle connect event also: https://github.com/libp2p/js-libp2p/blob/5a62f4f22f3d9658ae7356568283d443a76df4ea/src/registrar.ts#L41

Reply:
This case has been handled.